gcc/testsuite/rust/compile/builtin_macro_eager4.rs

```diff
@@ -0,0 +1,22 @@
+#![feature(rustc_attrs)]
+
+#[rustc_builtin_macro]
+macro_rules! cfg {
+  () => {{}};
+}
+
+macro_rules! a{
+  () => {"test"};
+}
+
+
+fn main () {
+  let os = "linux";
+  cfg!(); // { dg-error "cfg-pattern required" }
+  cfg!(1); // { dg-error "expected identifier" }
+  cfg!(windows); // shoud not error
+  cfg!(target_os = "linux"); //should not error
+  cfg!(a!()); // {dg-error "expected 1 cfg-pattern"}
```

Comment:
Maybe add one more test to make sure we get proper cfg-patterns out of macro expansion?
```suggestion
  cfg!(a!()); // {dg-error "expected 1 cfg-pattern"}
  cfg!(b!()); // ok - expands to `linux`
```

Reply:
I have trouble figuring out what is the propery way to return a cfg-pattern by simple Macro. I tried a few things but none of them work. Could you please enlight me?
Here is my attempts;
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e7bc6709ee61ba4d8e6beb03cd87fa71

Reply:
I tried multiple things including something like `cfg!(target_os = linux!())`, and that is not allowed either. So I'm not sure why `cfg!()` is a macro which should be expanded eagerly. @flip1995 or @bjorn3 you might have an idea?

Reply:
Why would eager expansion be necessary for `cfg!`? Grepping through the `rustc` source code I think the only things that are eagerly expanded are

- `concat[_*]!`
- `option_env!` and `env!`
- `compiler_error!`
- `include[_*]!`
- Something in `proc_macro_server.rs`. I haven't looked into what this does.

Reply:
I grepped through `rustc` for `MacEager`, which gave a match in the `expand_cfg` function. What pattern did you search for?

Reply:
I grepped for `fully_expand_fragment` that does the actual expansion of an item and is used (according to the [rustc def guide](https://rustc-dev-guide.rust-lang.org/macro-expansion.html?highlight=eager#eager-expansion)) to do eager expansion. And then I traced back the functions that calls that to where they are used.

`expand_cfg` doesn't call that function AFAICT. It calls `parse_meta_item` which is IIUC the same thing you can give it when using it as an attribute. So if you have the code that parses the `cfg` attribute, you should be able to just use that to parse the `cfg!` macro.

Reply:
Aaaah, nice, thank you @flip1995! In that case sorry @ArberSephirotheca for misguiding you and giving you a non-issue :/ if you'd like to work on something else, we can discuss it together on Zulip? Really sorry and I hope you didn't spend too much time on this